pkg/pillar/types/physicalioadapters.go

```diff
@@ -43,10 +43,7 @@ type PhysicalIOAdapter struct {
 	Assigngrp    string
 	Usage        zcommon.PhyIoMemberUsage
 	UsagePolicy  PhyIOUsagePolicy
-	// FIXME: cbattr - This needs to be thought through to be made into
-	//  a structure OR may be even various attributes in PhysicalIO structure
-	// itself.
-	// map <string, string> cbattr = 8;
+	Cbattr       map[string]string
```

Comment:
Note that with a map you will get a random order each time you populate the map. That might make it harder to look for changes.

Reply:
And please review the functions which operate on PhysicalIOAdapter to see if there is a copy or a comparision needed for the new Cbattr.

Reply:
Here for map comparison we can use reflect package's [DeepEqual](https://go.dev/src/reflect/deepequal.go?s=2432:2441#L98) function

Reply:
Didn't know that is supposed to work because I've run into issues where slices might not compare as equal even though the content inside the slices was the same, but I can't reproduce that right now. In any case, I think it is easier to have specific fields in the golang struct(s) in EVE-OS as I said above.

Reply:
No, you're right, having fields in struct makes it easier to debug and work with, including map makes sense if we to have volatile structure and we don't want to struggle with API changes. In wiki [proposal](https://wiki.lfedge.org/display/EVE/SR-IOV+Support) we used `cbattr` field to extend API, it was previously commented. I don't remember why we used it, maybe @milan-zededa can help.

Reply:
The main reason why I wanted to reuse `cbattr` (apart from the fact there it is already there), is that this is inside `PhysicalIO` structure, which is defined for any IO, not just NICs. All existing attributes in that structure are pretty generic (such as usage, assign group, etc.) and common to all IO types. And in the case of physical addresses, where values will differ between different IO types, we use the map `phyaddrs`.
So I'm open to introducing a field for SR-IOV attributes, but we need to design it carefully, such that it is extensible beyond NICs.
And regarding the issues with the map - we already have there the aforementioned `phyaddrs`, so I would just replicate how it is treated (if we decide to stay with `cbattr`).

Reply:
@eriknordmark @milan-zededa For SR-IOV we won't need more fields. Because SR-IOV is needed to create `n` VFs from PF on NIC. It's used combined with CPU-pinning and PCI-passthrough to provide high performance networking, both CPU-pinning and PCI-passthrough are different functionalities with their own APIs.

As for GPU SR-IOV I don't think it's not in year scope, since
- technology itself (GPU SR-IOV) is not mature yet
- drivers as of now are closed source
- in order to use it, virtualisation license from Nvidia is required

So I think if we define future as having SR-IOV on specified NIC (like Intel) with PCI-passthrough (and maybe CPU-pinning) with some tools to measure its' performance,  we will be okay. For this future one field `sriov_numvfs` in API is enough. Having `sriov_numvfs` in every `PhysicalIOAdapter` is fine, that way it'll be readable, although it's very flexible to use maps as it's done in `phyaddrs` [here](https://github.com/lf-edge/eve/blob/c1a2e0ffae0b68f9d655eee296945fa7a8474833/pkg/pillar/cmd/zedagent/parseconfig.go#L1225-L1244)